### ISSUE TITLE
DbContextOptionsBuilder<TContext>.UseLoggerFactory leads to memory leak

### ISSUE THREAD

**Issue by Dennis-Petrov** [dotnet/efcore#19146]
if application turns on logging via `DbContextOptionsBuilder<TContext>.UseLoggerFactory`, EF doesn't use model caching, and tries to build model again and again, each time context is created.

If context is configured with fluent API like this:
``` C#
        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            modelBuilder.HasDefaultSchema("dbo");
            modelBuilder.ApplyConfiguration(new PersonConfig());
        }
```
every call to context's constructor calls `OnModelCreating`, and, in turn, creates new instances of configuration types. These instances gets into static dictionary deep inside EF, and, ultimately, this dictionary occupies all available memory.

### Steps to reproduce

Just run this console app and hit breakpoint inside `OnModelCreating`:
``` C#
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;
using Microsoft.Extensions.Logging;
using System;
using System.Threading.Tasks;

namespace EFCore_MemoryLeak
{
    class Program
    {
        static void Main(string[] args)
        {
            var _ = TestContextAsync();

            Console.ReadLine();
        }

        static async Task TestContextAsync()
        {
            try
            {
                for (var i = 0; i < 10; i++)
                {
                    var optionsBuilder = new DbContextOptionsBuilder<SampleContext>();
                    var loggerProvider = new ConsoleLoggerProvider();
                    var loggerFactory = new LoggerFactory();

                    loggerFactory.AddProvider(loggerProvider);
                    optionsBuilder.UseLoggerFactory(loggerFactory);
                    optionsBuilder.UseNpgsql("User ID=postgres;Password=password;Host=localhost;Port=5432;Database=Foo_Db");

                    using (var context = new SampleContext(optionsBuilder.Options))
                    {
                        var people = await context.Set<Person>().ToListAsync();

                        Console.WriteLine(people.Count);

                        await Task.Delay(1000);
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }
        }
    }

    public class Person
    {
        public int Id { get; set; }

        public string Name { get; set; }
    }

    internal sealed class SampleContext : DbContext
    {
        public SampleContext(DbContextOptions options)
            : base(options)
        {
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            modelBuilder.HasDefaultSchema("dbo");
            modelBuilder.ApplyConfiguration(new PersonConfig());
        }
    }

    internal sealed class PersonConfig : IEntityTypeConfiguration<Person>
    {
        public void Configure(EntityTypeBuilder<Person> builder)
        {
            builder.HasKey(_ => _.Id);

            builder.Property(_ => _.Id)
                .ValueGeneratedOnAdd();

            builder.Property(_ => _.Name);

            builder.ToTable("People");
        }
    }

    internal sealed class ConsoleLoggerProvider : ILoggerProvider
    {
        public ILogger CreateLogger(string categoryName) => new ConsoleLogger();

        public void Dispose()
        {
        }
    }

    internal sealed class ConsoleLogger : ILogger
    {
        public IDisposable BeginScope<TState>(TState state) => new TraceLoggerScope();

        public bool IsEnabled(LogLevel logLevel) => true;

        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter) => Console.WriteLine(eventId);
    }

    internal sealed class TraceLoggerScope : IDisposable
    {
        public void Dispose()
        {
        }
    }
}
```

### Further technical details

EF Core version: 2.2.6
Database provider: Microsoft.EntityFrameworkCore.SqlServer, Npgsql.EntityFrameworkCore.PostgreSQL
Target framework: .NET Core 2.2
Operating system: Win10
IDE: Visual Studio 2019 16.3.9

**Comment by roji** [comment]
Take a look at [the documentation page on logging](https://docs.microsoft.com/en-us/ef/core/miscellaneous/logging), and specifically the following warning there:

> It is very important that applications do not create a new ILoggerFactory instance for each context instance. Doing so will result in a memory leak and poor performance.

In other words, instantiate your logger factory once and use that everywhere. You should be doing the same with the context options: build them once at the start of your program, and pass the same instance to all contexts.

**Comment by Dennis-Petrov** [comment]
OMG.

In EF6 times it wasn't perfect. We hoped that many things will be improved in EF Core without degrading existing developing experience. Working with EF Core I think from time to time, that it was designed by aliens from Mars or somewhere else.

Why on Earth to bind supporting functionality (and logging *is* supporting one) with model caching, which is one the basic things in EF? What for?

This is rhetorical question. Let it be here for history.

**Comment by ajcvickers** [comment]
@Dennis-Petrov It isn't bound directly to model caching, and this is fixed in 3.0 anyway.
